### PR TITLE
Implementation of subscription and unsubscription endpoints through websockets

### PR DESCRIPTION
**Pull request by bibekdahal**
Tests included

### CODE REVIEW

**Review by eoglethorpe — DISMISSED** [review]
single method for tests is OK but tests folder would still be good if we're having multiple test files

**Comment by bibekdahal** [comment]
We should not have a common tests folder, in the same way, we should not have a common models, views or serializers folder. Having tests inside the directory of the related app or feature makes the project more manageable once it starts growing.

However, we can have a tests folder inside each feature or app in case such feature has multiple files.

I have done this for deep, redis_store and geo folders since those are the only ones affected by this feature branch. The rest of the folders will be changed in other respective branches.

**Comment by bibekdahal** [comment]
Rebased with develop due to conflict with src/public directory.

**Comment by bibekdahal** [comment]
If there is no issue with the code related to this particular feature (subscription and unsubscription through websockets), can we merge this pull request?

I suggest we keep the discussion of common matters like above to issues.

**Comment by eoglethorpe** [comment]
Yep - I agree no global test folders but good to have within each module

**Comment by bibekdahal** [comment]
@eoglethorpe I don't know if there was a way for you to approve the change request you created. I was just playing around and dismissed it. Hope it's okay.

Also, can anybody do the latest review and merge the pull request if possible? Some new features will depend on code in this pull request. @thenav56 ?

**Comment by thenav56** [comment]
@tnagorra since you have worked on websocket before, please review and merge this request.

**Comment by eoglethorpe** [comment]
Ok I’m not sure but I’ll check tomorrow or Monday.

File: backend/deep/subscription.py [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+                    # Then remove the group from the redis list for the client
+                    Group(code).discard(self.message.reply_channel)
+                    r.lrem(redis_key, 0, code)
+
+                    # If empty, remove the redis list
+                    if r.llen(redis_key) == 0:
+                        r.delete(redis_key)
+
+                    # Reply back with the unsubscribed_codes.
+                    self.send({'ubsubscribed_codes': [code], 'success': True})
```
Comment (tnagorra) [inline backend/deep/subscription.py]: Is "ubsubscribed_codes" intentional or a typo?

File: backend/deep/subscription.py [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+                        unsubscribed_codes.append(code)
+
+                        code = r.lpop(redis_key)
+
+                    # If the list is empty, we may as well remove the list
+                    # from the redis store.
+                    if r.llen(redis_key) == 0:
+                        r.delete(redis_key)
+
+                    # Reply back with the unscubscribed_codes
```
Comment (tnagorra) [inline backend/deep/subscription.py:103]: Maybe a typo here:
Reply back with the **unscubscribed_codes**

File: backend/deep/subscription.py (hunk 1 above)
Comment (bibekdahal) [reply backend/deep/subscription.py]: typo :laughing:

**Pull request merged**